Bug 1741144: Update styles for dashboard cards

rawagner opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1741144

notable changes: 
- Dashboard card body has padding set to 0
- all lines (inventory, utilization, top consumers) are full body width (thanks to the change above)
- if padding is required, it needs to be specified case by case, Im using PF's `--pf-c-card--child--Padding*` variables
- I've added custom style to utilization chart empty state to have consistent height of row for empty and loaded charts

@andybraren can you please take a look at let me know if I need to fix anything else ?
![screencapture-localhost-9000-dashboards-2019-08-15-12_28_39](https://user-images.githubusercontent.com/2078045/63089606-afdc7a00-bf58-11e9-8d59-feab330613e5.png)

![screencapture-localhost-9000-dashboards-2019-08-15-12_39_06](https://user-images.githubusercontent.com/2078045/63089901-c6370580-bf59-11e9-8481-ccd6a8b110bd.png)


@cloudbehl can you also check storage and noobaa with loaded data ? I've checked error states of those dashboards and it looks fine.

rawagner commented on frontend/public/components/dashboard/dashboard-card/card.scss:
```diff
@@ -2,8 +2,8 @@
 
-.co-dashboard-card__body {
-  margin-top: 1.5em;
+.pf-c-card__body.co-dashboard-card__body {
```
I wasnt able to override the default padding by passing `co-dashboard-card__body` because `pf-c-card__body` style took precedence over mine. It seems like Im missing some configuration somewhere, but couldnt figure it out

rhamilto replied:
It seems like this will be a common use case that PatternFly should support.  I recommend opening an upstream issue.

Rather than overriding PatternFly here (which can lead to a maintenance issue for us since it changes the default behavior), I recommend using padding and negative margins that match the `.co-dashboard-card__body` padding values on `.co-utilization-card__item` (see https://getbootstrap.com/docs/3.3/css/#grid for an example).  This way, the change is an explicit opt-in that we control, and we are not changing the behavior of the upstream, thus requiring an addition to every default `.co-dashboard-card__body`.

rhamilto replied:
Edited above comment to change `.co-inventory-card__item` to `.co-*-card__item` as these classes are card-specific.  Probably want a generic class here instead to control add negative margins and padding.

andybraren replied:
@mcoker @matthewcarleton is there a recommended approach here?

mcoker replied:
+1 to [opening an issue in PF](https://github.com/patternfly/patternfly-next/issues/new) for this support. As a temporary workaround, we do have a new experimental divider component you could use and define in between card sections. There is no padding on the card itself, so the divider would span the width of the card as long as it's not in one of the header, body, or footer sections. I've created a codepen to show one way of using it in the card to create section dividers - https://codepen.io/mcoker/pen/WNewzGg

rawagner replied:
thank you all for suggestions, I went with negative margin approach as the divider does not go well with how we structure our components in dashboard. Issue reported https://github.com/patternfly/patternfly-next/issues/2153

rawagner commented:
@andybraren also as @spadgett  mentioned in https://github.com/openshift/console/pull/2330#issuecomment-520949954 should I change the size of icons for alerts ?

spadgett commented:
/assign @rhamilto

andybraren commented:
@rawagner For 4.2 let’s reduce the size of the Health icon to match the size of the alert and event icons. One size everywhere.

And thank you for this PR @rawagner, I was excited to see this in my inbox 🙂.

I’m currently away from a computer so I can’t get specific CSS changes for you, but here are a few things I notice:

The space between the icon and the text in the adjusted EmptyChart of Utilization is probably too large - is that something we can easily change?

The space below the Capacity donuts could also be reduced to match the space above them. Same for the space below the bars of Top Consumers.

The space around the charts of Utilization in general could be reduced to more closely align with the mockups if possible (condensed). I know adjusting the styles of those charts is trickier though.

cloudbehl commented:
@rawagner ceph dashboard also looks good to me. @afreen23 can you check the noobaa dashboard once. 
![Screenshot from 2019-08-16 17-01-43](https://user-images.githubusercontent.com/6695156/63164941-95cd9500-c047-11e9-92ad-82ac050237da.png)
![Screenshot from 2019-08-16 17-01-36](https://user-images.githubusercontent.com/6695156/63164942-96662b80-c047-11e9-9dbf-5ede0e03636f.png)
![Screenshot from 2019-08-16 17-01-10](https://user-images.githubusercontent.com/6695156/63164943-96fec200-c047-11e9-84aa-c20c73acdaf0.png)
![Screenshot from 2019-08-16 17-00-36](https://user-images.githubusercontent.com/6695156/63164945-96fec200-c047-11e9-8512-cea6f73ad366.png)

spadgett commented:
@cloudbehl The icon and text in the chart empty state are too close to the border in the screenshot you pasted. Have we changed the empty state style?

@andybraren What do you think of removing the gray background from the empty state? I'm not sure what PF4 recommends. I believe it draws attention to the empty charts instead of the charts with data and is a different style than the other empty state messages in console.

spadgett commented:
I'd suggest having 1, 2, or 4 columns in the cluster capacity card depending on screen width. 3 on top, 1 on bottom looks unbalanced and doesn't save any space. 2x2 would be better.

rawagner commented:
@andybraren styles updated.

![screencapture-localhost-9000-dashboards-2019-08-16-14_33_01](https://user-images.githubusercontent.com/2078045/63168125-8f352280-c033-11e9-8496-d2f8804879c2.png)
![screencapture-localhost-9000-dashboards-2019-08-16-14_37_17](https://user-images.githubusercontent.com/2078045/63168129-90fee600-c033-11e9-9d1f-fcf92f95a6d4.png)


> The space between the icon and the text in the adjusted EmptyChart of Utilization is probably too large - is that something we can easily change?
done. @spadgett this is what you are talking too, right ? EmptyChart in Utilization card.


>The space below the Capacity donuts could also be reduced to match the space above them. Same for the space below the bars of Top Consumers.

removed default margin added by console's chart components.

> The space around the charts of Utilization in general could be reduced to more closely align with the mockups if possible (condensed). I know adjusting the styles of those charts is trickier though.

should be better now

spadgett commented:
Yes, the icon and text are too close to the edge and the icon looks too large for the space to my eye. It might be fine if we remove the gray background, which I think is distracting anyway. I'd like @andybraren's opinion, though.

rawagner commented:
> @andybraren What do you think of removing the gray background from the empty state? I'm not sure what PF4 recommends. I believe it draws attention to the empty charts instead of the charts with data and is a different style than the other empty state messages in console.

@spadgett we are using console's empty states - `No datapoints found` is the message, see https://github.com/openshift/console/blob/master/frontend/public/components/graphs/graph-empty.tsx#L13.

spadgett commented:
I'm fine with the message itself. But there's not enough padding for icon and text that large.

spadgett commented:
To be clear, I’m proposing removing the gray background from all chart empty states in console. It’s inconsistent with other console empty state messages

matthewcarleton commented:
+1 on removing the background on the empty state. The PF pattern does not have this.

cloudbehl commented:
@spadgett I agree, the image and the text in the empty state is too close. We can remove the background image. It looks much better in that way.

![Screenshot from 2019-08-16 19-06-58](https://user-images.githubusercontent.com/6695156/63171418-29f42800-c059-11e9-9ac9-b114eade3aa1.png)
![Screenshot from 2019-08-16 19-06-52](https://user-images.githubusercontent.com/6695156/63171420-2a8cbe80-c059-11e9-8ad3-ec44bba4b442.png)

andybraren commented:
+1 to the background removal for all the reasons @spadgett described above, looks much better and more aligned with PF guidelines as @matthewcarleton said.

2x2 for the Capacity card would also be great, probably with some containing divs. This is looking really great @rawagner. 🙂

And I agree @cloudbehl the font size for the “Not datapoints found” should be the same across all cards, likely the smaller text (I’m away from my computer at the moment, but it should probably be the same size as most of the other font sizes we’re using in these cards).

afreen23 commented:
@cloudbehl all looks good except unavailable state of Resource Provider
![respro](https://user-images.githubusercontent.com/25664409/63173016-99b7e200-c05c-11e9-97bc-505030c7a62b.png)

@rawagner ^

rawagner commented:
Removed background and fixed font size in empty state

![screencapture-localhost-9000-dashboards-2019-08-16-16_01_32](https://user-images.githubusercontent.com/2078045/63173441-3324cb00-c040-11e9-8684-0c4e80751ac1.png)

![screencapture-localhost-9000-dashboards-persistent-storage-2019-08-16-16_01_17](https://user-images.githubusercontent.com/2078045/63173444-3455f800-c040-11e9-8ead-5378906805a0.png)

TODO: 2x2 in capacity, unavailable state in Resource Providers

andybraren commented:
@matthewcarleton @mceledonia What do you think about `var(--pf-c-empty-state__icon--Color)` in the images above? The icons seem pretty dark to me and still draw a lot of attention. Maybe we should change them to `#ededed` (the lightest color of the Capacity card's outer ring), `#d2d2d2` (second lightest), or `#bbb` (darkest)?

spadgett commented:
Honestly, I wonder if we even need the icon. IMO it draws focus from the charts that actually have data.

andybraren commented:
Here's `#d2d2d2`. Size of the Resiliency card's icon aside, I think keeping the icons but changing the color is reasonable.

![image](https://user-images.githubusercontent.com/9122899/63175726-94ce4080-c012-11e9-9c98-99dfe8d6f779.png)

rhamilto commented:
> Honestly, I wonder if we even need the icon. IMO it draws focus from the charts that actually have data.

+1

spadgett commented:
Looks good.

Not new, but the Data Resiliency icon is bigger than the other empty state icons.

matthewcarleton commented:
@andybraren I can't recall exactly but I think the darker colour is for a11y. I'm not sure we'd need to support contrast for an icon in an empty state though.

rawagner commented:
I noticed that on latest master empty state is overflowing the container 
![empty_overflow](https://user-images.githubusercontent.com/2078045/63341920-8fdafb00-c34a-11e9-8e74-45e80de1bf2e.png)

this is cause by changes in https://github.com/openshift/console/pull/2366 . I've removed the fixed height @TheRealJon can you take a look ?

I've also tried to set the color of icon in empty state to `#d2d2d2` but Im once again having issue with styles from PF overriding my own - also reported here https://github.com/patternfly/patternfly-react/issues/2694#issuecomment-522888136 

Are we going to keep the icons or remove them ? @andybraren @matthewcarleton

matthewcarleton commented:
@rawagner I'd agree, there should never be a fixed height applied to cards.
I'd also like to know more about why you can't override the styles. This should be possible.
We are also meant to remove the grey background on the empty state correct? 
From the PatternFly docs:
>Icon/image (optional): Display associated object icons here. An icon is only recommended for full page empty states
So I'd say we can remove.

spadgett commented:
cc @sg00dwin as he was also looking at the empty state overflowing its container on the project page

mceledonia commented:
Those icons are definitely too dark and don't require a11y contrast ratios, so a lighter color like you posted above would work well. I'd also be ok with not having an icon, we built the empty state component to be flexible to allow for any combination of elements (icon, text, button, etc.)

sg00dwin commented:
> cc @sg00dwin as he was also looking at the empty state overflowing its container on the project page

Yep, will be opening up a pr today with a fix for that. 
issue https://jira.coreos.com/browse/CONSOLE-1690

rawagner commented:
Removed icon from empty state and also added `text-secondary` style to empty state message. Optinions ?
![empty](https://user-images.githubusercontent.com/2078045/63416653-1998d000-c400-11e9-9ce8-17b4861e73b5.png)
![empty1](https://user-images.githubusercontent.com/2078045/63416657-1b629380-c400-11e9-9621-4e8ca76c9fb6.png)

yes, data resiliency is still using the exclamation mark as that card is not using Empty state, do we want to change that ? @andybraren

andybraren commented:
> yes, data resiliency is still using the exclamation mark as that card is not using Empty state, do we want to change that ? @andybraren

@rawagner Yeah that card is a weird one. Maybe we can just change the color of that icon to `#d2d2d2` and be happy for now until that whole card goes away in the future. 🙂 Otherwise we'd have something like "No datapoints available. Not available" which is weird.

Thank you @rawagner!

rawagner commented:
> > yes, data resiliency is still using the exclamation mark as that card is not using Empty state, do we want to change that ? @andybraren
> 
> @rawagner Yeah that card is a weird one. Maybe we can just change the color of that icon to `#d2d2d2` and be happy for now until that whole card goes away in the future.  Otherwise we'd have something like "No datapoints available. Not available" which is weird.
> 
> Thank you @rawagner!

I meant to replace icon & text with Empty state. Done in latest update

![resiliency_no_data](https://user-images.githubusercontent.com/2078045/63506087-115e9480-c4d5-11e9-9f0a-06119cabb204.png)

rawagner commented:
@andybraren @spadgett please review

spadgett commented:
@rhamilto ptal

rawagner commented:
rebased. @rhamilto

rawagner commented:
@rhamilto ok, Im now reusing the rules for things like card with no padding, icon sizes and text sizes. Please take a look

rawagner commented:
/retest

rawagner commented:
@rhamilto please add lgtm/approve labels if you are ok with PR.

mareklibra commented:
/lgtm
